toolkit/tools/pkg/imagecustomizerlib/customizeutils.go

```diff
@@ -348,33 +350,289 @@ func enableOrDisableServices(services imagecustomizerapi.Services, imageChroot *
+	if err != nil {
+		return err
+	}
+
+	// Batch process module to disable
+	err = ensureModulesDisabled(modulesToDisable, moduleLoadFilePath)
+	if err != nil {
+		return err
+	}
+
+	// Batch process module options
+	aggregatedOptions := []string{}
+	if len(moduleOptionsUpdates) > 0 {
+		for moduleName, options := range moduleOptionsUpdates {
+			for key, value := range options {
+				logger.Log.Infof("Writing module options %s=%s for module %s", key, value, moduleName)
+				aggregatedOptions, err = aggregateModuleOptions(aggregatedOptions, moduleOptionsFilePath, moduleName, key, value)
+				if err != nil {
+					return fmt.Errorf("failed to append module option for module %s: %w", moduleName, err)
+				}
+			}
+		}
+
+		file, err := os.OpenFile(moduleOptionsFilePath, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0644)
+		if err != nil {
+			return fmt.Errorf("failed to open %s: %w", moduleOptionsFilePath, err)
+		}
+		defer file.Close()
+
+		content := strings.Join(aggregatedOptions, "\n") + "\n"
+		// Append the content to the file.
+		if _, err := file.WriteString(content); err != nil {
+			return fmt.Errorf("failed to write module options to %s: %w", moduleOptionsFilePath, err)
+		}
+	}
+
+	return nil
+}
+
+func ensureModulesLoaded(moduleNames []string, moduleLoadFilePath string) error {
+	content, err := os.ReadFile(moduleLoadFilePath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("failed to read module load configuration: %w", err)
+		}
+		// If the file does not exist, initialize content as empty
+		content = []byte{}
+	}
+
+	needUpdate := false
+
+	for _, moduleName := range moduleNames {
+		if !strings.Contains(string(content), moduleName+"\n") {
+			content = append(content, (moduleName + "\n")...)
+			needUpdate = true
+			logger.Log.Infof("Loading module %s", moduleName)
+		} else {
+			logger.Log.Infof("Module %s is already loaded", moduleName)
+		}
+	}
+
+	if needUpdate {
+		err = os.WriteFile(moduleLoadFilePath, content, 0644)
+		if err != nil {
+			return fmt.Errorf("failed to update module load configuration: %w", err)
+		}
+	}
+	return nil
+}
 
-	for _, module := range modules.Load {
-		logger.Log.Infof("Loading kernel module (%s)", module.Name)
-		moduleFileName := module.Name + ".conf"
-		moduleFilePath := filepath.Join(imageChroot.RootDir(), "/etc/modules-load.d/", moduleFileName)
-		err = file.Write(module.Name, moduleFilePath)
+func ensureModulesDisabled(moduleNames []string, moduleDisableFilePath string) error {
+	content, err := os.ReadFile(moduleDisableFilePath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("failed to read disable configuration: %w", err)
+		}
+		// If the file does not exist, initialize content as empty
+		content = []byte{}
+	}
+
+	contentString := string(content)
+	needUpdate := false
+
+	for _, moduleName := range moduleNames {
+		blacklistEntry := "blacklist " + moduleName
+		if !strings.Contains(string(content), blacklistEntry+"\n") {
```

Comment:
It seems odd that you have 2 different ways of checking if a module is disabled: this `if` check, and the code in `isModuleDisabled`.

Also, this code seems simpler than the code in `isModuleDisabled`.